server/src/test/java/io/crate/integrationtests/SnapshotRestoreIntegrationTest.java

```diff
@@ -976,6 +982,61 @@ public void test_restore_partitioned_tables_rename_all() throws Exception {
         );
     }
 
+    @Test
+    public void test_restore_old_users() {
+        execute(
+            "CREATE REPOSITORY users_repo TYPE \"fs\" with (location=?, compress=true, readonly=true)",
+            new Object[]{USERS_REPO.getAbsolutePath()}
+        );
+        execute("CREATE USER \"John\" WITH (password='johns-password')");
+        execute("CREATE USER \"Arthur\"");
+        execute("CREATE USER \"Marios\"");
+        execute("SELECT * FROM sys.users ORDER BY name");
+        assertThat(response).hasRows(
+            "Arthur| NULL| false",
+            "John| ********| false",
+            "Marios| NULL| false",
+            "crate| NULL| true");
+
+        // Snapshot contains the following users:
+        // CREATE USER "Arthur" WITH (password='arthurs-password');
+        // CREATE USER "Ford" WITH (password='fords-password');
+        // CREATE USER "John";
+        execute("RESTORE SNAPSHOT users_repo.usersnap USERS with (wait_for_completion=true)");
+
+        execute("SELECT * FROM sys.users ORDER BY name");
+        assertThat(response).hasRows(
+            "Arthur| ********| false",
+            "Ford| ********| false",
+            "John| NULL| false",
+            "crate| NULL| true");
+
+        // Before any CREATE/ALTER/DROP operation RolesMetadata still have the users&roles defined
+        // but only UsersMetadata are used
+        RolesMetadata rolesMetadata = cluster().clusterService().state().metadata().custom(RolesMetadata.TYPE);
+        assertThat(rolesMetadata).isNotNull();
+        Assertions.assertThat(rolesMetadata.roles()).containsOnlyKeys("Arthur", "John", "Marios");
+        UsersMetadata usersMetadata = cluster().clusterService().state().metadata().custom(UsersMetadata.TYPE);
+        assertThat(usersMetadata).isNotNull();
+        assertThat(usersMetadata.users()).containsOnlyKeys("Arthur", "Ford", "John");
+
+        execute("ALTER USER \"Arthur\" SET (password='arthurs-new-password')");
```

Comment:
Do you think I should test the other two actions individually (CREATE USER/ROLE, and DROP USER/ROLE) to verify that this is the point we write `RolesMetadata` and we drop `UsersMetadata` ?

Reply:
Yep, thx!